stylix: test the global enable option using assertions

danth opened the pull request:
This adds a pair of assertions to each testbed: that the system is not changed in any way when Stylix is disabled, and that it is changed when Stylix is enabled.

I originally wanted this to compare the value of every option, but that uncovered a variety of errors in upstream code which is not normally evaluated - things like `meta.maintainers` definitions using undefined variables. I realised it would also require filtering out certain subtrees, such as the stylix.enable option itself, which we expect to change.

In the end, I opted to just compare the toplevel derivations instead.

## Things done

<!--
Please check what applies. Note that these are not hard requirements but merely
serve as information for reviewers.
-->
- [x] Tested locally
- [ ] Tested in [testbed](https://stylix.danth.me/testbeds.html)
- [x] Commit message follows [commit convention](https://stylix.danth.me/commit_convention.html)
- [ ] Fits [style guide](https://stylix.danth.me/styling.html)

### Maintainer CC

<!---
If you are editing an existing target, consider pinging relevant
`modules/<module>/meta.nix` module maintainers.
-->

danth committed:
stylix: test the global enable option using assertions

This adds a pair of assertions to each testbed: that the system is not
changed in any way when Stylix is disabled, and that it is changed when
Stylix is enabled.

I originally wanted this to compare the value of every option, but that
uncovered a variety of errors in upstream code which is not normally
evaluated - things like `meta.maintainers` definitions using undefined
variables. I realised it would also require filtering out certain
subtrees, such as the stylix.enable option itself, which we expect to
change.

In the end, I opted to just compare the toplevel derivations instead.

danth commented:
One thing to note is that the "stylix should affect the system when enabled" assertion doesn't know which module that effect is coming from: it's not testing that the module related to the testbed does something, but rather that *any* module does something.

We could potentially improve this by adding an extra system configuration, where the module in question is added to `disabledModules`, and then ensure something changes compared to the normal enabled configuration.

trueNAHO commented on stylix/testbed.nix:
```diff
@@ -146,23 +146,42 @@ let
+          ];
+        };
+
+        disabled = systems.clean.extendModules {
+          modules = [
+            inputs.self.nixosModules.stylix
+          ];
+        };
+
+        enabled = systems.disabled.extendModules {
+          modules = [
+            { inherit stylix; }
+          ];
+        };
```
What about removing the `enable = true;` attribute from:

https://github.com/danth/stylix/blob/2fb8321ea16c595e0208b22021ddaf1f471c634a/stylix/testbed.nix#L206-L241

Then, we define the `disabled` variant with `enable = false;` and the `enabled` variant with `enable = true;`?

danth replied:
The churrent implementation should also fail if there are any "option _ must be set" errors when having no Stylix configuration- as we want all features to be optional this is another useful check.

Perhaps this should be documented in a comment.

danth committed:
ci: limit to one evaluation worker

Testbeds need a lot of memory to evaluate, so following the number
of cores is not the best idea.

I originally tried to scale based on the amount of memory available,
but this is platform dependent so didn't work on MacOS. It's also not
the best idea since other processes (such as our build jobs) may
allocate memory after the value was read.

I think it makes the most sense to just use one evaluation worker;
`nix flake check` does this anyway but by using `nix-fast-build` we
are still gaining the ability to skip builds which are in the binary
cache. (Eventually, this may become a feature in Nix itself.)

danth committed:
stylix: mock `stylix.targets` options in `clean` testbed

Some testbeds need to configure options for their target, for example
the Firefox testbed which enables Firefox GNOME theme. This fails when
Stylix is not installed (which we use to determine the original state
of the system, to see if Stylix changed anything). We can't just
exclude the testbed configuration from the clean system, because then
it would always change due to the target software being added.

danth commented:
Not sure if this is actually a good idea, as it triples the cost of evaluating each testbed. We also need to mock out certain options to be able to import the testbed file while not importing Stylix.

trueNAHO commented:
For reference, this PR would partially automate the process that would resolve https://github.com/danth/stylix/issues/543#issuecomment-2323517576. It is "partially automate" because not all targets and options are covered by testbeds.

> Not sure if this is actually a good idea, as it triples the cost of evaluating each testbed. We also need to mock out certain options to be able to import the testbed file while not importing Stylix.

What about introducing a new testbed that composes all target testbeds, with the following variants:

- clean: Not importing Stylix
- enabled: `enable = true;`
- disabled: `enable = false;`

The heuristic for selecting testbeds would be to prefer `/modules/<MODULE>/testbeds/default.nix` over any other `/modules/<MODULE>/testbeds/<TESTBED>.nix` testbed. If there is no `/modules/<MODULE>/testbeds/default.nix`, then the testbed could be chosen by something arbitrary like:

```nix
builtins.head (
  lib.naturalSort (
    builtins.attrNames (builtins.readDir /* /modules/<MODULE>/testbeds */)
  )
)
```

danth commented:
The main issue is that the `clean` copy doesn't have any of the `stylix.*` options defined, and some testbeds contain things like `stylix.targets.firefox.firefoxGnomeTheme.enable = true` which then fails. To add mock versions of the options, we have to introduce a fourth copy since doing this to `clean` would conflict with the real options when we define them.
